optaplanner-core/src/main/java/org/optaplanner/core/api/score/holder/AbstractScoreHolder.java

```diff
@@ -67,16 +68,44 @@ public boolean isConstraintMatchEnabled() {
     // Worker methods
     // ************************************************************************
 
+    protected class MultiLevelConstraintUndoListener implements ActivationUnMatchListener {
+
+        private Map<Integer, ActivationUnMatchListener> activationUnMatchListenerMap;
+
+        public MultiLevelConstraintUndoListener(int scoreLevel, ActivationUnMatchListener activationUnMatchListener) {
+            this.activationUnMatchListenerMap = new HashMap<Integer, ActivationUnMatchListener>();
```

Comment:
Few ideas:
* A: Initialize capacity statically to 1 or 2 (most score types have two levels, and 99% of rules will only need 1 level (backwards compatibility))
* B: Initialize capacity to `scoreLevel` (might be good in the average case, although is a bit "pulled out of a hat")
* C: Somehow set it according to the score type to an exact (maximal) size for the given type